src/wp-includes/class-wp-textdomain-registry.php

```diff
@@ -56,6 +68,12 @@ public function get( $domain, $locale ) {
 			return $this->all[ $domain ][ $locale ];
 		}
 
+		// If no path is found for the given locale, check if a path has been set
+		// using load_plugin_textdomain/load_theme_textdomain.
+		if ( 'en_US' !== $locale && isset( $this->default_path[ $domain ] ) ) {
```

Comment:
Note that, upon further testing on my side, the problem is actually _not_ limited to only `en_US`, it's just most relevant there. The bug occurs whenever translation files are not present (see my ticket comment as well).

Without knowing the code too well, I would assume there's something off with the original commit, where the code assumes that a language file must be present locally in order to set it.

Specifically, that seems to be line 762 in `l10n.php`: By adding `$wp_textdomain_registry->set( $domain, $locale, false );` before the early `return false;`, the problem appears to be fixed.

Reply:
As per the ticket, this does not work as it breaks locale switching tests.